src/transformers/models/zamba2/modular_zamba2.py

```diff
@@ -0,0 +1,1872 @@
+    attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim/2)
+    Finally, this attention layer contributes to tied transformer blocks aimed to increasing compute without increasing model size. Because this
+    layer is tied, un-tied adapters (formally the same as LoRA but used in the base model) modules are added to the q, k, v projectors to increase
+    expressivity with a small memory overhead (see Fig. 2 of https://arxiv.org/pdf/2411.15242).
+    """
+
+    def __init__(
+        self,
+        config: Zamba2Config,
+        layer_idx: Optional[int] = None,
+        num_fwd_mem_blocks: int = None,
+        block_id: int = None,
+    ):
+        super().__init__(config, layer_idx)
+        self.num_fwd_mem_blocks = num_fwd_mem_blocks
+        self.layer_block_map = layer_type_list(config)
+        self.block_id = block_id
+        self.is_causal = True
+
+        if config.use_shared_attention_adapter:
+            self.linear_q_adapter_list = nn.ModuleList([])
+            self.linear_k_adapter_list = nn.ModuleList([])
+            self.linear_v_adapter_list = nn.ModuleList([])
+
+            for i in range(self.num_fwd_mem_blocks):
+                if i % config.num_mem_blocks == block_id:
+                    linear_q_adapter = nn.Sequential(
+                        nn.Linear(self.attention_hidden_size, self.config.adapter_rank, bias=False),
+                        nn.Linear(self.config.adapter_rank, self.attention_hidden_size, bias=False),
+                    )
+                    linear_k_adapter = nn.Sequential(
+                        nn.Linear(self.attention_hidden_size, self.config.adapter_rank, bias=False),
+                        nn.Linear(self.config.adapter_rank, self.attention_hidden_size, bias=False),
+                    )
+                    linear_v_adapter = nn.Sequential(
+                        nn.Linear(self.attention_hidden_size, self.config.adapter_rank, bias=False),
+                        nn.Linear(self.config.adapter_rank, self.attention_hidden_size, bias=False),
+                    )
+                else:
+                    linear_q_adapter = nn.Identity()
+                    linear_k_adapter = nn.Identity()
+                    linear_v_adapter = nn.Identity()
+                self.linear_q_adapter_list.append(linear_q_adapter)
+                self.linear_k_adapter_list.append(linear_k_adapter)
+                self.linear_v_adapter_list.append(linear_v_adapter)
+
+        if config.use_mem_rope:
+            rope_theta = config.rope_theta
+            if config.use_long_context:
+                a = 8
+                rope_theta = rope_theta * a ** (config.attention_head_dim / (config.attention_head_dim - 2))
+            self.rotary_emb = Zamba2RotaryEmbedding(
+                config.attention_head_dim,
+                max_position_embeddings=config.max_position_embeddings,
+                base=rope_theta,
+            )
```

Comment:
This should be removed (defined in the Model), and `position_embeddings` should be passed everywhere as a required arg (not a keyword arg)!

Reply:
Removed [here](https://github.com/huggingface/transformers/pull/34517/commits/99708af8505771e07d77cee20fcc7c533e562861). We cannot make `position_embeddings` required as some zamba2 models do not have rope.

I also updated rope [here](https://github.com/huggingface/transformers/pull/34517/commits/d9b4a5004af2f53e281038c66ee4f984bb53dafe#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL382). I modified the init method compared to the inherited rope as we need to pass custom `dim` and `base` arguments.